### PR TITLE
Correctly handle auto downsampling in codec

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

1. Handle `max_source_resolution=auto` correctly.
2. Generating cache key based on resolution level instead of time

### CODE REVIEW

File: pkg/queryfrontend/cache_splitter.go [hunk 1]
```diff
@@ -8,17 +8,32 @@ import (
...
+
+func newThanosCacheSplitter(interval time.Duration) thanosCacheSplitter {
+	return thanosCacheSplitter{
+		interval:    interval,
+		resolutions: []int64{downsample.ResLevel2, downsample.ResLevel1, downsample.ResLevel0},
+	}
+}
 
 // TODO(yeya24): Add other request params as request key.
 // GenerateCacheKey generates a cache key based on the Request and interval.
```
Comment (bwplotka) [inline pkg/queryfrontend/cache_splitter.go:29]: Can we update comment?

File: pkg/queryfrontend/cache_splitter.go [hunk 2]
```diff
@@ -8,17 +8,32 @@ import (
...
-func (t constSplitter) GenerateCacheKey(_ string, r queryrange.Request) string {
-	currentInterval := r.GetStart() / time.Duration(t).Milliseconds()
+func (t thanosCacheSplitter) GenerateCacheKey(_ string, r queryrange.Request) string {
```
Comment (bwplotka) [inline pkg/queryfrontend/cache_splitter.go:30]: It's not splitter anymore, rather `cacheKeyGenerator` ? (:

File: pkg/queryfrontend/cache_splitter_test.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/thanos-io/thanos/pkg/testutil"
+)
+
+func TestGenerateCacheKey(t *testing.T) {
+	splitter := newThanosCacheSplitter(hour)
+
+	for _, tc := range []struct {
+		name             string
+		req              queryrange.Request
+		expectedCacheKey string
```
Comment (bwplotka) [inline pkg/queryfrontend/cache_splitter_test.go:20]:
```suggestion
		expected string
```

File: pkg/queryfrontend/cache_splitter.go [hunk 4]
```diff
@@ -8,17 +8,32 @@ import (
...
+	currentInterval := r.GetStart() / t.interval.Milliseconds()
 	if tr, ok := r.(*ThanosRequest); ok {
-		return fmt.Sprintf("%s:%d:%d:%d", tr.Query, tr.Step, currentInterval, tr.MaxSourceResolution)
+		i := 0
+		for ; i < len(t.resolutions) && t.resolutions[i] > tr.MaxSourceResolution; i++ {
```
Comment (bwplotka) [inline pkg/queryfrontend/cache_splitter.go:34]: I think that's good for start :muscle:

File: pkg/queryfrontend/roundtrip_test.go [hunk 5]
```diff
@@ -300,6 +300,16 @@ func TestRoundTripCacheMiddleware(t *testing.T) {
 		MaxSourceResolution: 10 * seconds,
 	}
 
+	// same query params as testRequest, different maxSourceResolution
```
Comment (bwplotka) [inline pkg/queryfrontend/roundtrip_test.go:303]: Can we start comments with upper case letter like normal sentence?

File: pkg/queryfrontend/roundtrip_test.go [hunk 6]
```diff
@@ -345,13 +355,18 @@ func TestRoundTripCacheMiddleware(t *testing.T) {
...
+		{
+			name:     "different max source resolution but still same level",
 			req:      testRequest3,
 			expected: 3,
 		},
 		{
-			name:     "do it again",
-			req:      testRequest2,
+			name:     "different max source resolution and different level",
+			req:      testRequest4,
```
Comment (bwplotka) [inline pkg/queryfrontend/roundtrip_test.go:369]: why not inlining those requests maybe?

**Review by bwplotka — COMMENTED** [review]
Some suggestions only. LGTM! (:

File: pkg/queryfrontend/cache_splitter.go (hunk 1 above)
Comment (yeya24) [reply pkg/queryfrontend/cache_splitter.go:29]: Sorry, but what should be updated here?

File: pkg/queryfrontend/cache_splitter.go (hunk 2 above)
Comment (yeya24) [reply pkg/queryfrontend/cache_splitter.go:30]: Yes, that's better

**Review by bwplotka — APPROVED** [review]
:muscle: LGTM!

Thanks!

**Pull request merged**